includes/class-wc-admin-reports-coupons-stats-segmenting.php

```diff
@@ -0,0 +1,312 @@
+			"SELECT
+						MAX($table_name.date_created) AS datetime_anchor,
+						{$intervals_query['select_clause']} AS time_interval,
+						$segmenting_groupby
+						$segmenting_select
+					FROM
+						$table_name
+						$segmenting_from
+						{$intervals_query['from_clause']}
+					WHERE
+						1=1
+						{$intervals_query['where_time_clause']}
+						{$intervals_query['where_clause']}
+						$segmenting_where
+					GROUP BY
+						time_interval, $segmenting_groupby
+					$segmenting_limit",
+			ARRAY_A
+		); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+		// Reformat result.
+		$intervals_segments = $this->reformat_intervals_segments( $intervals_segments, $segmenting_groupby );
+		return $intervals_segments;
+	}
+
+	/**
+	 * Return array of segments formatted for REST response.
+	 *
+	 * @param string $type Type of segments to return--'totals' or 'intervals'.
+	 * @param array  $query_params SQL query parameter array.
+	 * @param string $table_name Name of main SQL table for the data store (used as basis for JOINS).
+	 *
+	 * @return array
+	 * @throws WC_REST_Exception In case of segmenting by variations, when no parent product is specified.
+	 */
+	protected function get_segments( $type, $query_params, $table_name ) {
+		global $wpdb;
+		if ( ! isset( $this->query_args['segmentby'] ) || '' === $this->query_args['segmentby'] ) {
+			return array();
+		}
+
+		$product_segmenting_table = $wpdb->prefix . 'wc_order_product_lookup';
+		$unique_orders_table      = '';
+		$segmenting_where         = '';
+
+		// Product, variation, and category are bound to product, so here product segmenting table is required,
+		// while coupon and customer are bound to order, so we don't need the extra JOIN for those.
+		// This also means that segment selections need to be calculated differently.
+		if ( 'product' === $this->query_args['segmentby'] ) {
+			$segmenting_selections     = array(
+				'product_level' => $this->get_segment_selections_product_level( $product_segmenting_table ),
+				'order_level'   => $this->get_segment_selections_order_level( $table_name ),
+			);
+			$segmenting_from           = "INNER JOIN $product_segmenting_table ON ($table_name.order_id = $product_segmenting_table.order_id)";
+			$segmenting_groupby        = $product_segmenting_table . '.product_id';
+			$segmenting_dimension_name = 'product_id';
+
+			$segments = $this->get_product_related_segments( $type, $segmenting_selections, $segmenting_from, $segmenting_where, $segmenting_groupby, $segmenting_dimension_name, $table_name, $query_params, $unique_orders_table );
+		} elseif ( 'variation' === $this->query_args['segmentby'] ) {
+			if ( ! isset( $this->query_args['product_includes'] ) || count( $this->query_args['product_includes'] ) !== 1 ) {
+				throw new WC_REST_Exception( 'woocommerce_rest_invalid_segmenting_variation', __( 'product_includes parameter need to specify exactly one product when segmenting by variation.', 'wc-admin' ), 400 );
```

Comment:
Should this validation take place in the REST controller logic instead of here - maybe a validation callback on the `segmentby` query arg? Just not sure if this logic will ever be invoked outside of the context of a REST request.

Reply:
Updated exception handling in https://github.com/woocommerce/wc-admin/pull/1286/commits/42fff45f447541f4569a6a7cf3f3dcbad42ef7ba.
As I believe it does not make much sense to compare variations from different products, I kept the condition, at least for now.